http/middleware.go

```diff
@@ -62,42 +64,35 @@ func processRequest(tx types.Transaction, req *http.Request) (*types.Interruptio
 		// body inspection, otherwise we just let the request follow its
 		// regular flow.
 		if req.Body != nil && req.Body != http.NoBody {
-			_, err := io.Copy(tx.RequestBodyWriter(), req.Body)
+			b, err := io.ReadAll(req.Body)
```

Comment:
Isn't there not much difference since we read the entire body into memory here?

I had an idea to process the limited body without copying everything here think it would be better

https://github.com/corazawaf/coraza/pull/539#discussion_r1062020076

Reply:
I guess what you are implying here is

```
			_, err := io.CopyN(tx.RequestBodyWriter(), req.Body, limit)
			if err != nil {
				return nil, err
			}

			reader, err := tx.RequestBodyReader()
			if err != nil {
				return nil, err
			}
			// Add any remaining bytes beyond the coraza limit to its buffer
			reader = io.MultiReader(reader, req.Body)
			// req.Body is transparently reinizialied with a new io.ReadCloser.
			// The http handler will be able to read it.
			// Prior to Go 1.19 NopCloser does not implement WriterTo if the reader implements it.
			// - https://github.com/golang/go/issues/51566
			// - https://tip.golang.org/doc/go1.19#minor_library_changes
			// This avoid errors like "failed to process request: malformed chunked encoding" when
			// using io.Copy.
			// In Go 1.19 we just do `req.Body = io.NopCloser(reader)`
			if rwt, ok := reader.(io.WriterTo); ok {
				req.Body = struct {
					io.Reader
					io.WriterTo
					io.Closer
				}{reader, rwt, req.Body}
			} else {
				req.Body = struct {
					io.Reader
					io.Closer
				}{reader, req.Body}
			}
```

While I think this is good (brilliant indeed) in proxy-wasm we still copy the body (although limited) instead of just looking at the body size (way much simpler because we have that number already).

I agree, doing ReadAll isn't ideal, however I tried `Copy` in the past but failed miserably to fix the Multipart test (exactly https://github.com/corazawaf/coraza/blob/v3/dev/http/middleware_test.go#L78). If you have an idea on how to fix this I would love to hear it.

Regarding the `ProcessBodySize` method, I think it is better that the transaction itself rejects or continues the flow if the body limit condition is met rather than replicating the logic in all connectors.